fdbserver/MasterProxyServer.actor.cpp

```diff
@@ -1913,6 +1913,18 @@ ACTOR Future<Void> ddMetricsRequestServer(MasterProxyInterface proxy, Reference<
 	}
 }
 
+ACTOR static Future<Void> debugReadTxnStateStoreServer(RequestStream<DebugReadTxnStateStoreRequest> requests,
```

Comment:
A proxy from an old generation can still reply to this request, so I think you'll eventually find a simulation failure that looks like:

1. The previous generation of proxies are partitioned away from the master/cluster controller
2. More commits are done on the next generation of proxies.
3. CycleTest verification starts running.
4. The client sends requests to the old generation of proxies, and sees stale data
5. Half way through, the old proxies die / the client gets the ServerDBInfo update with the new generation of proxies
6. the client sends the second half of the requests to the new generation of proxies, and sees recent data
7. The test fails with cycletest reporting corruption

I think you'd need to do the same "no TLogs are locked" check that happens before responding to GRVs.

Reply:
The cycle test only does one read from the txnStateStore currently, and it's at the end after confirming that all cycle clients have stopped and there are no in-flight transactions. The test also waits until it's reading the txnStateStore at a version that's higher than the commit version of the transaction that killed any in-flight transactions. I think this is sufficient to avoid false positives.

That said, this debugReadTxnStateStoreServer thing does have some weird semantics - you can read uncommitted data and it doesn't provide causal consistency as you pointed out.

Reply:
I missed the uncommitted data part.  Oof.  Your added comment is 👍